### ISSUE TITLE
Excessively large type for `maxItems`

### ISSUE THREAD

**Issue by Nevon** [bcherny/json-schema-to-typescript#372]
Given a property such as:

```json
{
  "type": "array",
  "minItems": 0,
  "maxItems": 1000,
  "items": {
    "type": "string",
    "anyOf": [
      {
        "$ref": "#/definitions/ipv4CidrBlockpattern"
      }
    ],
    "not": {
      "enum": [
        "0.0.0.0/0"
      ]
    }
  }
}
```

The generated type will be:

```ts
type ips = []
  | [Ipv4CidrBlockpattern & string]
  | [Ipv4CidrBlockpattern & string, Ipv4CidrBlockpattern & string]
  | [Ipv4CidrBlockpattern & string, Ipv4CidrBlockpattern & string, Ipv4CidrBlockpattern & string]
# ... and so on for each possible length of the array
```

Given a `maxItems` of 1000, the type definition becomes _truly_ massive, and even fails to be generated in some cases.

Given a `TupleOf` type ([from here](https://github.com/microsoft/TypeScript/issues/26223)) we could shrink this down quite a bit:

```ts
type BuildPowersOf2LengthArrays<N extends number, R extends never[][]> =
  R[0][N] extends never ? R : BuildPowersOf2LengthArrays<N, [[...R[0], ...R[0]], ...R]>;

type ConcatLargestUntilDone<N extends number, R extends never[][], B extends never[]> =
  B["length"] extends N ? B : [...R[0], ...B][N] extends never
    ? ConcatLargestUntilDone<N, R extends [R[0], ...infer U] ? U extends never[][] ? U : never : never, B>
    : ConcatLargestUntilDone<N, R extends [R[0], ...infer U] ? U extends never[][] ? U : never : never, [...R[0], ...B]>;

type Replace<R extends any[], T> = { [K in keyof R]: T }

type TupleOf<T, N extends number> = number extends N ? T[] : {
  [K in N]: 
  BuildPowersOf2LengthArrays<K, [[never]]> extends infer U ? U extends never[][]
  ? Replace<ConcatLargestUntilDone<K, U, []>, T> : never : never;
}[N];

type ips = Tuple<Ipv4CidrBlockpattern & string, 0>
  | TupleOf<Ipv4CidrBlockpattern & string, 1>
  | TupleOf<Ipv4CidrBlockpattern & string, 2>
```

This will still generate 1000 tuples in the union, but each tuple will only contain the parameter type once, rather than N times.

**Comment by panstromek** [comment]
We also have this problem for even smalller arrays but with more complex objects. Also, the generated type cannot be easily named/referenced and its items become distinct types. I'd prefer to just have the generated type be simple array and leave out the length completely, otherwise you get into the dependant type territory too much which just brings unnecessary amounts of coplexity, while it's not needed for most cases, because the length is runtime value anyway.

**Comment by Nevon** [comment]
While it's not documented in the cli options, you can actually disable the min/maxlength check with the `--ignoreMinAndMaxItems` option. This is what we're doing now.

**Comment by panstromek** [comment]
That's good to know, thanks ;)

**Comment by bcherny** [comment]
`ignoreMinAndMaxItems` is documented in the README (https://github.com/bcherny/json-schema-to-typescript#options). Let's also document it in CLI options.

**Comment by bcherny** [comment]
Fixed in a89ffe1d2158ecdb3f0859a395b52450767089e6

**Comment by bcherny** [comment]
Fixed in a89ffe1d2158ecdb3f0859a395b52450767089e6
